02_bubble/marksomething/main.go

```diff
@@ -0,0 +1,101 @@
+package main
+
+import (
+	"fmt"
+	"io"
+	"os"
+)
+
+var out io.Writer = os.Stdout
+
+func bubble(s []int) []int {
+
+	// Copy slice: See https://github.com/go101/go101/wiki
+	sorted := append(s[:0:0], s...)
+
+	for range sorted {
+		noSwaps := true
+
+		// need to do n-1 compare/swaps
+		for i := range sorted[1:] {
+
+			if sorted[i] > sorted[i+1] {
+				sorted[i], sorted[i+1] = sorted[i+1], sorted[i]
+				noSwaps = false
+			}
+		}
+		if noSwaps {
+			break
+		}
+
+	}
+	return sorted
+}
+
+func insertion(s []int) []int {
+
+	//pre-allocate output, maintain high water mark in loop
+	sorted := make([]int, len(s))
+
+	for hwm, needle := range s {
+
+		//find insertion point
+		insPoint := hwm
+		for i, v := range sorted[:hwm] {
+			if needle < v {
+				insPoint = i
+				break
+			}
+		}
+
+		//move data at and after insertion point up by 1 and insert new value
+		copy(sorted[insPoint+1:hwm+1], sorted[insPoint:hwm])
+		sorted[insPoint] = needle
+	}
+	return sorted
+}
+
+func quick(s []int) []int {
+
+	// Quicksorting inplace, so make a copy
+	// Copy slice: See https://github.com/go101/go101/wiki
+	sorted := append(s[:0:0], s...)
+
+	// Need to predefine so we can recurse
+	var qsort func([]int, int, int)
+	qsort = func(slc []int, low int, high int) {
+
+		// no value
+		if low == high {
+			return
+		}
+		//single value
+		if low+1 == high {
```

Comment:
maybe pull the last to if's together?
`if low >= high`  or `if low==high ||l ow+1 == high`

Reply:
Good point, have used `if low==high || low+1 == high`
note: that's not the same as `if low >= high`